lib/server.js

```diff
@@ -208,15 +208,27 @@ function emitRouteError(server, req, res, err) {
 
     req.log.trace({name: name, err: err}, 'entering emitRouteError');
 
-    if (server.listeners(name).length > 0) {
-        server.emit(name, req, res, err, once(function () {
-            res.send(err);
-            server._finishReqResCycle(req, res, null, err);
-        }));
-    } else {
+    // Check to see if any listeners are registered for the error, if not go
+    // ahead and finish the request. If there are listeners, emit the error and
+    // wait for all of the handlers to finish before finishing the request.
+    function cleanup() {
         res.send(err);
         server._finishReqResCycle(req, res, null, err);
     }
+
+    var invoked = 0;
+    function cb() {
+        // Check to see if we have received as many callbacks as there are
+        // listeners
+        if (invoked++ === server.listenerCount(name)) {
```

Comment:
I think this is a breaking change, i.e., the response won't get flushed up until all listener callbacks are invoked. I don't have opinions about this opinionated cleanup (badum psh) but probably worth documenting in the migration guide.